### ISSUE TITLE
Question about more idiomatic fsharp option check for exactlyOneOrDefault

### ISSUE THREAD

**Issue by giuliohome** [fsprojects/SQLProvider#483]
### Description

This is a follow up to issue [#478](https://github.com/fsprojects/SQLProvider/issues/478)

### Repro steps

Steps required to reproduce the problem

1. Step A
Do everything like for the above mentioned issue #478

2. Step B
Then try to write a more functional/idiomatic code so change this code

```fsharp
        let foundPrice = 
             query {
             for p in prices do 
             where (p.Provider = aPrice.Provider  && p.RefDate = Some aPrice.RefDate)
             exactlyOneOrDefault
             } 
         
        if not(isNull foundPrice) then
            foundPrice.MyValue <- Some (aPrice.Price + 1.1)
          else 
            let newRec = prices.Create()  
            newRec.RefDate <- Some aPrice.RefDate
            newRec.Provider <- aPrice.Provider
            newRec.MyValue <- Some aPrice.Price
```

to a more idiomatic

```fsharp
        let foundPriceMaybe = 
             query {
             for p in prices do 
             where (p.Provider = aPrice.Provider  && p.RefDate = Some aPrice.RefDate)
             select (Some p)
             exactlyOneOrDefault
             } 
         
        match foundPriceMaybe with
        |Some foundPrice ->
            foundPrice.MyValue <- Some (aPrice.Price + 1.1)
        |None ->
            let newRec = prices.Create()  
            newRec.RefDate <- Some aPrice.RefDate
            newRec.Provider <- aPrice.Provider
            newRec.MyValue <- Some aPrice.Price
```

### Expected behavior

No error, same behavior, just a nicer source code

### Actual behavior

I get a runtime exception when I submit the context updates  

```fsharp
System.Exception was unhandled
Message: An unhandled exception of type 'System.Exception' occurred in FSharp.Data.SqlProvider.dll
Additional information: Error - you cannot update an entity that does not have a primary key. (main.very_urgent)
```

### Known workarounds

I'm forced to stay with the ugly null check

### Related information 

Same as previous issue except of course latest package 1.1.19

**Comment by gaston-e-nusimovich** [comment]
The null check, while not as elegant, is effective.

**Comment by Thorium** [comment]
Meanwhile I use myself:
```fsharp
async {
   let! foundPriceMaybe = 
       query {
           for p in prices do 
           where (p.Provider = aPrice.Provider  && p.RefDate = Some aPrice.RefDate)
           select (p)
       } |> Seq.tryHeadAsync
   match ...
}
```

...just a mention that the error you got `(you cannot update an entity that does not have a primary key)` is not about the code but about your table:

You have to have a primary key in your table if you do `foundPrice.MyValue <- Some (aPrice.Price + 1.1)` because we will generate sql `update x where id = ...` and now if you don't have any primary key of any kind, we cannot say which item should we update. The SQL clauses are totally separate, there is no way to say "the one that I did have in the previous query", the database data might even have changed during your other logics.

**Comment by giuliohome** [comment]
Quick remark: *of course* there is the primary key already detailed in the DDL of the repro for previous issue. This is only an ORM package problem, there is no update in the table in the meanwhile... and btw there are transactions just in case one wants to enforce this check

**Comment by giuliohome** [comment]
@gaston-e-nusimovich no need to do the null check any longer

this better work-around is effective too and nicer ;-)
```fsharp
let foundPriceMaybe = 
     query {
     for p in prices do 
     where (p.Provider = aPrice.Provider  && p.RefDate = Some aPrice.RefDate)
     } |> Seq.tryHead
```

**Comment by giuliohome** [comment]
Anyway, I hope that also this bug can be fixed eventually.

Also, I'd love to see the `null` check in the CRUD example replaced with `| Some` pattern matching :-)

**Comment by Thorium** [comment]
we do accept PRs e.g. to improve the documentation ;-)

**Comment by giuliohome** [comment]
Ok, I could produce a PR ony reproduce the bug here...
Basically just look at the following two tests, the first (more complex) passes while the second (even simpler!) fails! It is very strange. I've tried to debug it but the code is not easy for me and there are runtime optimization so I can not even look inside the data dictionary...

```fhsarp

[<Test >]
let ``option from join select with exactly one``() =
    let dc = sql.GetDataContext()
    let qry = 
        query {
            for cust in dc.Main.Customers do
            join ord in dc.Main.Orders on (cust.CustomerId = ord.CustomerId)
            where (cust.CustomerId = "ALFKI" && ord.OrderId = (int64 10643))
            select (Some (cust, ord))
            exactlyOneOrDefault
        } 

    match qry with
    | Some(cust, ord) ->
        let id = cust.CustomerId
        let ordId = ord.OrderId
        Assert.AreEqual("ALFKI",id)
        Assert.AreEqual(10643,ordId)
    | None ->
        Assert.Fail()

[<Test >]
let ``option from simple select with exactly one``() =
    let dc = sql.GetDataContext()
    let qry = 
        query {
            for cust in dc.Main.Customers do
            where (cust.CustomerId = "ALFKI")
            select (Some cust)
            exactlyOneOrDefault
        } 

    match qry with
    | Some cust ->
        let id = cust.CustomerId
        Assert.AreEqual("ALFKI",id)
    | None ->
        Assert.Fail()
     
```

**Comment by Thorium** [comment]
I think you shouldn't need select (Some x) as you never select None anyways.

**Comment by giuliohome** [comment]
Oh that (Some x) is very funny :-) lol 

I also thought it was kind of wrong or useless... then I arrived to this explanation:
its purpose is to make the result set an option so that its default value is None => so, it is exactlyOneOrDefault which is using None.

At the point when I was full of doubt about the usage of (Some x), I found that the query from the join was working, so it is something in the ORM that simplify things when it finds only one alias.

Consider that adding only a fictitious join is enough to make it work

```fsharp
    let qry = 
        query {
            for cust in dc.Main.Customers do
            join cust2 in dc.Main.Customers on (cust.CustomerId=cust2.CustomerId) //ugly but enough to make it work
            where (cust.CustomerId = "ALFKI")
            select (Some cust)
            exactlyOneOrDefault
        } 
```

I really don't understand why yesterday I could debug the nuget tests and today after trying for hours I can't step into them !!! I'm try with the interactive method that you suggest later, thanks

**Comment by giuliohome** [comment]
I have a possible patch (if ok can be extended to all the other implementation) related to [this line](https://github.com/fsprojects/SQLProvider/blob/master/src/SQLProvider/Providers.SQLite.fs#L586) 

```fsharp
            let selectcolumns =
                if projectionColumns |> Seq.isEmpty then "1" else
                String.Join(",",
```

it gets executed as sql so a trivial fix could be simply

```fsharp
            let selectcolumns =
                if projectionColumns |> Seq.isEmpty then "*" else
                String.Join(",",
```

**Comment by giuliohome** [comment]
Found a safer change without any `select *`, just intercepting a parameter from the lambda in this line of [QueryExpression](https://github.com/fsprojects/SQLProvider/blob/master/src/SQLProvider/SqlRuntime.QueryExpression.fs#L59)

```fsharp
            | eOther -> 
                if eOther.NodeType.ToString().Contains("Parameter") then
                    let param = eOther :?> ParameterExpression
                    if (param.Type.Name.Equals("SqlEntity")) then
                        let alias = Utilities.resolveTuplePropertyName (param.Name) tupleIndex
                        if aliasEntityDict.ContainsKey(alias) then
                            Some (alias,aliasEntityDict.[alias].FullName, None)
                        elif ultimateChild.IsSome then
                            Some (alias,fst(ultimateChild.Value), None)
                        else None
                    else None
                else None
```

**Comment by Thorium** [comment]
Merged to 1.1.20

**Comment by giuliohome** [comment]
Thank you!
Yes, now the 2 unit tests work as expected!
I assumed that would have solved also the original problem *in the first post here above* and I didn't check it. Unfortunately I was wrong and that update is still failing.
If you have time to look at it, I would appreciate it so much... I will try to do the same but it is a bit more difficult for me. I'm still using the workaround
